backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/predicate/impl/LuceneNumericTermsPredicate.java

```diff
@@ -0,0 +1,108 @@
+import java.util.List;
+
+import org.hibernate.search.backend.lucene.search.impl.AbstractLuceneCodecAwareSearchValueFieldQueryElementFactory;
+import org.hibernate.search.backend.lucene.search.impl.LuceneSearchContext;
+import org.hibernate.search.backend.lucene.search.impl.LuceneSearchValueFieldContext;
+import org.hibernate.search.backend.lucene.search.predicate.impl.AbstractLuceneLeafSingleFieldPredicate;
+import org.hibernate.search.backend.lucene.types.codec.impl.AbstractLuceneNumericFieldCodec;
+import org.hibernate.search.engine.search.common.ValueConvert;
+import org.hibernate.search.engine.search.predicate.SearchPredicate;
+import org.hibernate.search.engine.search.predicate.spi.TermsPredicateBuilder;
+
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.ConstantScoreQuery;
+import org.apache.lucene.search.Query;
+
+public class LuceneNumericTermsPredicate extends AbstractLuceneLeafSingleFieldPredicate {
+
+	private LuceneNumericTermsPredicate(Builder builder) {
+		super( builder );
+	}
+
+	public static class Factory<F, E extends Number>
+			extends
+			AbstractLuceneCodecAwareSearchValueFieldQueryElementFactory<TermsPredicateBuilder, F, AbstractLuceneNumericFieldCodec<F, E>> {
+		public Factory(AbstractLuceneNumericFieldCodec<F, E> codec) {
+			super( codec );
+		}
+
+		@Override
+		public Builder<F, E> create(LuceneSearchContext searchContext, LuceneSearchValueFieldContext<F> field) {
+			return new Builder<>( codec, searchContext, field );
+		}
+	}
+
+	private static class Builder<F, E extends Number> extends AbstractBuilder<F> implements TermsPredicateBuilder {
+		private final AbstractLuceneNumericFieldCodec<F, E> codec;
+
+		private E term;
+		private List<E> terms;
+		private boolean allMatch;
+
+		private Builder(AbstractLuceneNumericFieldCodec<F, E> codec, LuceneSearchContext searchContext,
+				LuceneSearchValueFieldContext<F> field) {
+			super( searchContext, field );
+			this.codec = codec;
+		}
+
+		@Override
+		public void matchingAny(Collection<?> terms, ValueConvert convert) {
+			allMatch = false;
+			fillTerms( terms, convert );
+		}
+
+		@Override
+		public void matchingAll(Collection<?> terms, ValueConvert convert) {
+			allMatch = true;
+			fillTerms( terms, convert );
+		}
+
+		@Override
+		public SearchPredicate build() {
+			return new LuceneNumericTermsPredicate( this );
+		}
+
+		@Override
+		protected Query buildQuery() {
+			if ( term != null ) {
+				return codec.getDomain().createExactQuery( absoluteFieldPath, term );
+			}
+
+			if ( !allMatch ) {
+				return codec.getDomain().createSetQuery( absoluteFieldPath, terms );
+			}
+
+			BooleanQuery.Builder builder = new BooleanQuery.Builder();
+			for ( E termItem : terms ) {
+				Query query = codec.getDomain().createExactQuery( absoluteFieldPath, termItem );
+				builder.add( query, BooleanClause.Occur.MUST );
+			}
+			return new ConstantScoreQuery( builder.build() );
```

Comment:
Same remark as for the Elasticsearch predicate:

> So if someone calls `constantScore()` in the DSL... we will wrap the query in a constant score query twice?
> 
> Could you do something similar to `org.hibernate.search.backend.lucene.types.predicate.impl.LuceneExistsPredicate.AbstractBuilder#AbstractBuilder` instead?
> 
> Additionally, you could also add a safeguard in `applyConstantScore()` that calls a new `isAlwaysConstant()` method, so that it doesn't wrap the query unnecessarily.